### ISSUE TITLE
Migrate docs.rs to RDS and ECS

### ISSUE THREAD

**Issue by jdno** [rust-lang/simpleinfra#353]

**Comment by syphar** [comment]
another thing we need to figure out: 
- how would the database migrations be run in the deploy process?

**Comment by syphar** [comment]
after checking our NGINX config there is a second piece we need to solve somehow: 

IP blocks. 

Every now and then we have a mis-acting crawler and in these cases we blocked the source IP in NGINX on our server. 

I would prefer to have this in AWS / CloudFront if possible. 

Otherwise we would add this to our web container, probably configured via environment variable?

**Comment by syphar** [comment]
Next piece we need before prod : 

Access to logs

**Comment by jdno** [comment]
For blocking IPs, we should just set up a web-application firewall ([AWS WAF](https://aws.amazon.com/waf/)). I actually think that we already have one set up for docs.rs, but I'm not 100% sure.

Access to the logs is a good point! It probably makes sense to stream all logs to a central place, whether that's CloudWatch or an external tool like Datadog.

**Comment by syphar** [comment]
One thought I had thinking about this topic again: 

- CloudFront has a hard limit on in-progress wildcard path invalidations (15)
- we are invalidating the crate docs after each build 

from https://github.com/rust-lang/docs.rs/issues/1871#issuecomment-1268744723 
> Looking at https://docs.rs/releases/activity it seems we average at least 600 releases per day. If an average invalidation takes 5 minutes and we can have 15 in parallel, that's 3 invalidations per minute throughput. With 1440 minutes in a day, we could handle up to 4320 builds per day before we wind up in unbounded growth land. Of course, that's based on a significant assumption about how long an invalidation takes.

I'm not sure if we can / should handle invalidations differently, but we might think about using fastly when we rework the infra?

**Comment by Mark-Simulacrum** [comment]
Can't we de-duplicate invalidations if we approach the limit? E.g., a * invalidation every 5 minutes would presumably never hit the limit. Not sure how that would affect cache hit rates, but I'd expect designing around not needing invalidations or being ok with fairly blanket invalidations to be a good long-term strategy.

(I think we've had this conversation elsewhere before).

**Comment by syphar** [comment]
> Can't we de-duplicate invalidations if we approach the limit? E.g., a * invalidation every 5 minutes would presumably never hit the limit.

You mean "escalating" them, so when the queue is too long, just convert the queue into a full purge. 
This is definitely would work, but would mean that the user experience (especially outside the US) is worse until the cache is fuller again. Of course this might be acceptible for us. 

> being ok with fairly blanket invalidations 

This also means that the backend always has to be capable to handle the full uncached load, and higher egress costs depending on how often we have to do the full purge. 

I also remember a discussion at EuroRust that we could think about having additional docs.rs webservers (also readonly DB & local bucket?) in some regions (europe?). 

>  I'd expect designing around not needing invalidations 

You're right, this is a valid discussion to have. I imagine this would only work when the URLs would include something like the build-number in the URL, and replace the more generic URLs rest with redirects. If I'm not missing something this would revert some of the SEO & URL work from https://github.com/rust-lang/docs.rs/issues/1438 (introducing `/latest/` URLs ). And then people would start linking specific docs-builds in their sites as they did before we had `/latest/`. 




> (I think we've had this conversation elsewhere before).

you're probably right :) 

I wanted to bring it up here as a point, for when we migrate infra anyways.

**Comment by Mark-Simulacrum** [comment]
Note that (IMO) if we can get the cache keys setup right, i.e. everything except HTML is always at a by-hash file path - it seems to me that /latest/ can just be served with a short ttl (5 minutes), perhaps with stale-while-revalidate. That means that there's a small period where it's not necessarily consistent what version you get from it across all pages if some are cached locally and some aren't (and likewise for CDN), but I don't see any real problem with that. Users mostly won't even notice.

Yes, especially anything out of s3 can definitely be replicated if we need it to be into multiple regions pretty easily. This just causes issues while you need invalidations since you're racing against replication which can itself take some time (hours IIRC for the cheap option and minutes for the costly one?)

**Comment by marcoieni** [comment]
Resuming this work after more than one year, so I'm creating a new task list.

We want to follow an incremental approach.

## Initial migration

This steps move the current docs.rs EC2 to the `docs-rs-staging` and `docs-rs-prod` accounts.

Benefits:
- the docs.rs team has a staging environment
- everything is under IaC

Notes:
- the important thing is that web server and database are always in the same region.

Tasks:

- [x] create the `docs-rs-prod` AWS account https://github.com/rust-lang/simpleinfra/pull/846
- [x] Give write access to the DB and S3 in staging to Docs-rs team. https://github.com/rust-lang/simpleinfra/pull/1018 (for the DB, we just give the connection string)
- [ ] For DataDog, also configure OTEL.
- [ ] implement Fastly CDN service for the staging environment. Give write permission to Denis
- [ ] implement rate limiting in fastly to mimic the nginx configuration we have in EC2
- [ ] migrate the DB in the EC2 to RDS. https://github.com/rust-lang/simpleinfra/pull/1002/
  - [x] Create the DB in the `docs-rs-staging` account. This DB will not be used initially.
  - [x] Create the DB in the `docs-rs-prod` account. Use a replica (2 instances). Regarding the size, set 100GB because `/var/lib/postgresql/` is 13GB in the EC2.
  - [ ] set read replica
  - [ ] make the EC2 database "readonly" by locking the build queue. This will stop registry watching, which is the only automated source for changes.
  - [ ] Wait for the cdn invalidator queue to get empty
  - [ ] backup the `docs-rs-prod` database from the EC2 database, and change the `DATABASE_URL` of the EC2 to point to the new database
  - [ ] destroy the DB in the EC2
  - [ ] share a `DATABASE_URL` to the docs.rs team securely so that they can make / debug changes. Note that a read-only aws role can't decrypt the item in the parameter store. Can they read this by ssh-ing into the machine?
- [ ] Migrate the webserver to ECS in both the staging and prod accounts. N instances
  - [ ] Datadog log shipping.
  - [ ] Datadog agent? Do we need to put it in the container?
  - [ ] Document how to do a rollback
  - [ ] Need local volume of 100 GB
- [ ] Migrate the registry watcher to ECS in both the staging and prod accounts. 1 instance.
  - [ ] Datadog log shipping
  - [ ] Need local volume of 100 GB
- [ ] Create the EC2 in the staging and prod account with terraform + ansible.
  - [x] Do we need to assign any IaM roles?
  - [x] provision the new EC2 via terraform
    - https://github.com/rust-lang/simpleinfra/pull/963
    - https://github.com/rust-lang/simpleinfra/pull/964
  - [x] copy the s3 bucket https://github.com/rust-lang/simpleinfra/pull/973/ https://github.com/rust-lang/simpleinfra/pull/998
  - [x] configure bastion https://github.com/rust-lang/simpleinfra/pull/982
  - [x] give docs-rs team permission to aws account
  - [ ] write and apply ansible
    - build the builder in CI and upload it in ECR 
    - docker compose to run the builder
  - [ ] delete the mechanism introduced in https://github.com/rust-lang/simpleinfra/pull/973/ and https://github.com/rust-lang/simpleinfra/pull/998 to copy the bucket and also the `west1` terragrunt alias in the account.json files
  - [ ] delete the old EC2 and the entire terraform module.
  - [ ] evaluate performance and change instance sizes of compute and DB accordingly
- [ ] give access to the crates-io-index-events queue to all components that should have access.
 
Old EC2 migration (optional):
  - [ ] take snapshot of the EC2
  - [ ] attach the snapshot to the new EC2
  - [ ] upgrade the new EC2 
  - [ ] If things went well, switch the dns to point to the new ec2
  - [ ] delete the old EC2 and the entire terraform module.

Milestones (staging):
- setup Fastly
- setup web server. N instances behind an aws load balancer hit by Fastly. Does this load balancer support rate limit?
- setup db from a backup (we have daily backups uploaded in S3)
- Increase the load incrementally from fastly and monitor resource
- deploy watcher

## After migration

After everything is under IaC, we can think about the following improvements to improve docs.rs and its DevX.

- [ ] Restore fastly shielding pop?
- [ ] scaling crates documentation build. this problem will get bigger. We could have an autoscaling group of EC2s.
  - we could scale based on the number of queued jobs, or simply with CPU as a start. Maybe we could use spot instances to save credits.
  - use packer?
- [ ] move to fastly as CDN mostly because of faster cache invalidation and hopefully better cache ratio. (there are two distributions: static (static.docs.rs pointing to s3) and main one (docs.rs)
  - [x] docs-rs team needs to implement the invalidations
  - [x] https://github.com/rust-lang/simpleinfra/pull/824/
  - [x] From Denis: would we reuse the current certificate? Or create a new one managed by fastly?
  - [ ] drop the hardcoded ip ranges and just use the [`Fastly-Client-IP` header](https://www.fastly.com/documentation/reference/http/http-headers/Fastly-Client-IP/) ([context](https://rust-lang.zulipchat.com/#narrow/channel/356853-t-docs-rs/topic/New.20docs.2Ers.20infrastructure/near/558522892)) 
  - [x] Create API key to purge both distributions if needed and give it to the team. (not sure if they need to store this securely somewhere eg in their password manager).
  - [ ] remove code duplication in `terraform/docs-rs/fastly-compute-docs-rs/bin/terraform-external-build.sh`
  - [x] https://github.com/rust-lang/simpleinfra/pull/711
  - [ ] add logs to datadog, similar to [crates-io](https://github.com/rust-lang/simpleinfra/blob/785cb526646f48218eac1cf80b13a0ac16214757/terragrunt/modules/crates-io/compute-static/src/main.rs#L89).
( when it's new, we could directly go wildcard)
  - [x] After completing this, close https://github.com/rust-lang/simpleinfra/issues/237 as it won't be relevant anymore.
- [ ] find a way to deal with scrapers. They do many requests and take down docs.rs Minor issue. Doesn't happen very often. Rate limiting? Probably fastly has tools for this
- [ ] Find a way to run DB migrations automatically Some ideas:
  - execute DB migrations from CI, before deploying all the other apps (assuming that the deployment is triggered from CI)
  - execute DB migrations from an app (either EC2 or one of the two ECS). Even if N replicas attempt to run the migration at the same time, postgres should handle the concurrency via locks or similar, right?
- [ ] have a subdomain per crate, to help with XSS, so this would have to be in the new CDN & DNS setup.
- [ ] monitor all VMs (bastion and EC2)

**Comment by marcoieni** [comment]
yes https://docs.aws.amazon.com/AmazonRDS/latest/PostgreSQLReleaseNotes/postgresql-extensions.html#postgresql-extensions-17x
